### PR TITLE
Adding a new FPGA tutorial: simple_host_streaming

### PR DESCRIPTION
**Pull request by tyoungsc**
Signed-off-by: tyoungsc <tanner.young-schultz@intel.com>

# Description
Adding a new tutorial for FPGAs simple_host_streaming

## Type of change
- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
- [X] Command Line

### CODE REVIEW

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+                << "[--chunks=<int>] "
+                << "[--chunk_count=<int>] "
+                << "[--inflight_kernels=<int>] "
+                << "[--iterations=<int>]\n";
+      return 0;
+    } else {
+      std::string str_after_equals = arg.substr(arg.find("=") + 1);
+
+      if (arg.find("--chunks=") == 0) {
+        chunks = atoi(str_after_equals.c_str());
```
Comment (keryell) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:91]:
```suggestion
        chunks = std::stoi(str_after_equals);
```
everywhere

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+  try {
+    // device selector
+#if defined(FPGA_EMULATOR)
+    INTEL::fpga_emulator_selector selector;
+#else
+    INTEL::fpga_selector selector;
+#endif
+
+    // queue properties to enable profiling
+    auto prop_list = property_list{ property::queue::enable_profiling() };
```
Comment (keryell) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:148]:
```suggestion
    property_list prop_list { property::queue::enable_profiling() };
```

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+
+    // create the device queue
+    queue q(selector, dpc_common::exception_handler, prop_list);
+
+    // make sure the device supports USM host allocations
+    device d = q.get_device();
+    if (!d.get_info<info::device::usm_host_allocations>()) {
+      std::cerr << "ERROR: The selected device does not support USM host"
+                << " allocations\n";
+      return 1;
```
Comment (keryell) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:158]: It is not clear why sometimes you `return 1` and sometimes you `std::terminate()`.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+    }
+
+    // the USM input and output data
+    Type *in, *out;
+    if ((in = malloc_host<Type>(total_count, q)) == NULL) {
```
Comment (keryell) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:163]:
```suggestion
    if ((in = malloc_host<Type>(total_count, q)) == nullptr) {
```
everywhere

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+    // generate the random input data
+    // NOTE: by generating all of the data ahead of time, we are essentially
+    // assuming that the producer of data (producing data for the FPGA to
+    // consume) has infinite bandwidth. However, if the producer of data cannot
+    // produce data faster than our FPGA can consume it, the CPU producer will
+    // bottleneck the total throughput of the design.
+    std::generate_n(in, total_count, [] { return Type(rand() % 100); });
+
+    // a lambda function to validate the results
+    auto validate_results = [&]() {
```
Comment (keryell) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:181]:
```suggestion
    auto validate_results = [&] {
```

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+      for (size_t i = 0; i < total_count; i++) {
+        auto comp = (in[i] == out[i]);
+        for (auto j = 0; j < comp.get_count(); j++) {
```
Comment (keryell) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:184]: I am lost here... What is the type of `comp`?

Comment (keryell) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:184]: I found it... `sycl::long8`.

**Review by keryell — CHANGES_REQUESTED** [review]
You could simplify a little bit the code.
Curious that all your figures are empty files.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp (hunk 1 above)
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:91]: Don't really feel like wrapping a try-catch around this.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp (hunk 3 above)
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:158]: ctrl-v ... :)

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp (hunk 6 above)
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:184]: I thought you liked `auto` ;)?

**Comment by tyoungsc** [comment]
> Curious that all your figures are empty files.

I don't see them as empty? I can click on "View Raw" and see them here.

**Comment by tyoungsc** [comment]
> > Curious that all your figures are empty files.
> 
> @keryell I don't see them as empty? I can click on "View Raw" and see them here.

Woops. Totally wrong. I'll fix that. Thanks!

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+// The following is a block diagram of this kernel:
+//
+// |-----------------|         |--------------------------|
+// | CPU    |-----|  | in_ptr  | FPGA  |---| InPipe       |
+// |        | RAM |--|---------|------>| P |========> ... |
+// |        |-----|  |         |       |---|              |
+// |-----------------|         |--------------------------|
+//
+template<typename T, typename InPipe>
+event SubmitProducer(queue &q, T* in_ptr, size_t size) {
```
[the remark this answers is not in the document]
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp:32]: Great minds :-).
I had this, I was asked nicely by a test FPGA user to add it explicitly.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+// |-----------------|         |--------------------------|
+//
+template<typename T, typename OutPipe>
+event SubmitConsumer(queue &q, T* out_ptr, size_t size) {
+  auto e = q.submit([&](handler& h) {
+    h.single_task<C>([=]() [[intel::kernel_args_restrict]] {
+      host_ptr<T> out(out_ptr);
+      for (size_t i = 0; i < size; i++) {
+        T data = OutPipe::read();
+        *(out + i) = data;
```
[the remark this answers is not in the document]
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp:64]: A bug in the front end prevents me from doing:
`out[i] = data;`
:-(.

**Comment by tyoungsc** [comment]
Weird. I had to `git rm` the PNG files, then re-add them to see any change.

**Review by akertesz — DISMISSED** [review]

**Review by JoeOster — APPROVED** [review]

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp (hunk 6 above)
Comment (keryell) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:184]: Indeed, I am a fan :-) but I was puzzled by not having `comp` being directly a `bool`...

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (keryell) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp:32]: At least could could have
```c++
template<typename InPipe, typename T = InPipe::value_type>
event SubmitProducer(queue &q, T* in_ptr, size_t size) {
```
to have lazy people happy while allowing some other people to work-out on their keyboard...

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (keryell) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp]: I spend my life saying pointers are badly unsafe and often not necessary in SYCL...
Fix the bug? :-) @bader

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp (hunk 1 above)
Comment (keryell) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/simple_host_streaming.cpp:91]: I see... I missed the
```c++
  // check the chunks
  if (chunks <= 0) {
    std::cerr << "ERROR: 'chunks' must be greater than 0\n";
    return 1;
  }
```
later.

But honestly a try catch is not that bad
```c++
try { chunks = std::stoi(str_after_equals); }
catch (...) {
  std::cerr << "ERROR: 'chunks' must be greater than 0\n";
  throw; // or std::terminate()
}
```

**Comment by keryell** [comment]
> Weird. I had to `git rm` the PNG files, then re-add them to see any change.

Not clear why... It looks more like a GitHub glitch.
Even now they are display with a size of `0` and `Empty file.` but if I click on it I can see the images!

**Review by tomlenth — APPROVED** [review]
looks great.  thank you

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp]: The issue is only when the pointer is:
1) written to
2) decorated with host_ptr or device_ptr. Something about operator[] returning a const?

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp:32]: I will suggest adding that to the pipe class (it currently doesn't have a `using value_type=T` line :(

[the remark this answers is not in the document]
Comment (keryell) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp:32]: https://github.com/triSYCL/triSYCL/blob/master/include/triSYCL/pipe.hpp#L65

[the remark this answers is not in the document]
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/simple_host_streaming/src/multi_kernel.hpp:32]: Lovely example. Great name; 'triSYCL'.

**Review by akertesz — APPROVED** [review]

**Pull request merged**